Prohibit IP addresses in ECHConfig.public_name. (#436)
tlswg/draft-ietf-tls-esni: draft-ietf-tls-esni.md

  RFC2119:
  RFC7918:

informative:

>   WhatWGURLAddressParser:

Reviewer: Perhaps s/WhatWGURLAddressParser/WHATWG-IPV4/ to be slightly shorter? Also, while not explicitly stated, the example in https://datatracker.ietf.org/doc/html/rfc7322#section-4.8.6.6 implies we like SHOUTY-CASE-WITH-HYPHENS for these tags.

Author: 👍 The variant in your PR is cleaner, so we can just use that if we decide to land this.

   title: "URL Living Standard - IPv4 Parser"

>    target: https://url.spec.whatwg.org/#concept-ipv4-parser

Author: Flagged as an open issue for now.

of a particular ECH configuration.

public_name
: The non-empty DNS name of the client-facing server, i.e., the entity trusted
to update the ECH configuration. This is used to correct misconfigured clients,
as described in {{handle-server-response}}. This value MUST NOT begin or end
with an ASCII dot and MUST be parsable as a dot-separated sequence of LDH
labels, as defined in {{!RFC5890}}, Section 2.3.1. Clients MUST ignore any
`ECHConfig` structure whose `public_name` does not meet these criteria. Note

> that these criteria allow IPv4 addresses in standard dotted-decimal or other

Reviewer 2: ```suggestion
that these criteria disallow IPv4 addresses in standard dotted-decimal or other
```
?

Author: Sadly, no :( The format admits IP addresses, but the following sentence says they should be ignored in those cases.

non-standard notations such as octal and hexadecimal (see {{?RFC3986}}, Section

> 7.4). Clients MUST ignore the `ECHConfig` if it contains an encoded IP address.

Reviewer 3: So this says "work out if it is an IP address and don't allow it".  That's too imprecise for me.

The [parsing algorithm for IP addresses in the whatWG URL spec](https://url.spec.whatwg.org/#concept-ipv4-parser) is very precise.  You could reference that and say that if that returns a value then the ECHConfig can be ignored.

A MUST here without that level of precision is not good.  I can't test this.  Well, I can make something up, but that's not how this works.

Author: Thanks -- I don't think any of us (beyond you) knew that reference existed. We can certainly cite it to make this algorithm more precise.

Author: @martinthomson I referenced it directly as a possible way to make this determination. I'm not sure we want to require use of this specific algorithm (especially since that spec is a living document and may change). We could alternatively try to pull that algorithm into this document, but that seems... terrible.

Reviewer: I agree with Martin that any MUST-level requirements should be precise. Though the WHATWG algorithm is... definitely hairy. :-/ I imagine most TLS libraries, including ours, will not have a version of it readily available, so we'd have to reimplement it.

I currently lean towards not trying too hard to filter IPs and instead saying:

* The public name is a DNS name. DNS names have such-and-such syntax. Anything outside that syntax is invalid. (We could, optionally, exclude the tighter [RFC3986 definition](https://datatracker.ietf.org/doc/html/rfc3986#section-3.2.2), but see note below.)
* When your application evaluates whether a certificate is valid for a public name, you must interpret the public name as a DNS name. That means, you should match it against DNS entries in the SAN list.
* Note, however, that the syntax for IPv4 literals in URLs are also valid DNS names. If your application would otherwise interpret the public name as an IP literal, you should make sure you don't go down the IP address matching logic.
* At the TLS level, we consider the embedding of IPv4 literals in strings an application-specific thing and assume different applications may do this subtly differently.

The consequence of this is:

* If your application's certificate logic already separates DNS strings and IP strings, you can just send the string down the DNS string path and don't worry too hard.
* If your application's certificate logic takes a string and expects to parse out DNS vs IP address out of one string for web stuff, you should detect this (hopefully now your application's IP parser is available)
* Servers that live in an ecosystem where "127.0.0.1" is not a valid DNS name should not use "127.0.0.1" as a public name because it won't work.

In other words, the web-specific `MakeOriginFromDNSName("https", "127.0.0.1", 443)` function should fail because `https://127.0.0.1` gives back an origin that's interpreted as an IP literal, not a DNS name. It's not possible to express "127.0.0.1", the DNS name in the web.

This is slightly disappointing because, practically, that WHATWG-IP-literals are invalid DNS names is really a universal constraint. But we seem to have made enough of a mess here that I don't think we can guarantee the application won't think it is an IPv4 literal. (RFC3986 has [another definition](https://datatracker.ietf.org/doc/html/rfc3986#section-3.2.2) of IPv4 literal, but if we used it, we'd still allow a bunch of strings that, on the web, is interpreted as an IPv4 literal. So applications would need to account for this `MakeOriginFromDNSName` error case regardless.)

Author: @davidben two quick comments:

1. If we say "anything outside that syntax is invalid," without giving a function to determine if it's invalid, then we're back at Martin's criticism, right?
2. If the check is deferred to the certificate check, doesn't this run the risk of including an IP address in the outer SNI extension?

Reviewer: > If we say "anything outside that syntax is invalid," without giving a function to determine if it's invalid, then we're back at Martin's criticism, right?

Sorry, when I say "that syntax", I mean just the syntax that DNS defines. The one that is sadly ambiguous with IPv4 literals.

> If the check is deferred to the application when validating the certificate, doesn't this run the risk of including an IP address in the outer SNI extension?

It does, but I think that's fine. I think the two risks here are that the DNS can induce the client to send some string in SNI to some server, and it technically violates RFC6066. The former doesn't haven't a clear threat model, more fuzzy "well random private servers are often buggy". I think it's reasonable to say we'll validate the DNS-level syntax and assume the server behaves equally well given "a.b.c.d" or "0x1.0x2.0x3.0x4". The latter is true, but RFC6066 doesn't actually define the syntax anyway. I'm not really sure how one is meant to interpret that prohibition apart beyond "look deep into your heart and if this *feels* like an IP address, don't do it". And under this interpretation, TLS truly,  deeply in its heart, wants a DNS name of "0x1.0x2.0x3.0x4". That the application will never accept any certificate for that *DNS* name is the application's problem. :-)

(This is not a *great* interpretation. It's just barely self-consistent enough to say with a straight face, in hopes of getting out of this hole.)

Author: Gotcha, thanks for clarifying! I'm fine with this as it avoids the parsing complexity altogether and moves it to other places in the stack -- "not my problem" =)

of a particular ECH configuration.

public_name

> : The non-empty DNS name of the client-facing server, i.e., the entity trusted

Reviewer: Couple minor nitpicks, just to make this less long-winded: I think we can drop "non-empty". It's already captured by both the struct definition and "DNS name". We can probably also replace "standard dotted-decimal or other non-standard notations such as octal and hexadecimal" with "dotted-decimal or other notations", since this PR ends up citing WHATWG later anyway. (Also it seems better not to get into the "standard" vs "non-standard" rabbithole.)

Author: 👍 I'm fine with this, too :)

to update the ECH configuration. This is used to correct misconfigured clients,
as described in {{handle-server-response}}. This value MUST NOT begin or end
with an ASCII dot and MUST be parsable as a dot-separated sequence of LDH
labels, as defined in {{!RFC5890}}, Section 2.3.1. Clients MUST ignore any
`ECHConfig` structure whose `public_name` does not meet these criteria. Note
that these criteria allow IPv4 addresses in standard dotted-decimal or other
non-standard notations such as octal and hexadecimal (see {{?RFC3986}}, Section
7.4). Clients SHOULD ignore the `ECHConfig` if it contains an encoded IPv4
address. (To determine if a public_name value is an IPv4 address, clients can
invoke the IPv4 parser algorithm in {{WhatWGURLAddressParser}}. It returns a

> value when the input is an IPv4 address.) Failure to validate that public_name

Reviewer 3: ```suggestion
value when the input is an IPv4 address.) Failure to validate that a public_name
```

> value when the input is an IPv4 address.) Failure to validate that a public_name
> is not an IPv4 address may cause connection failures to the client-facing
> server. However, it is not expected to induce security failures. Clients will
> fail closed if the client-facing server cannot produce a certificate bound
> to the reference identity based on the public_name.

Reviewer 4: This assertion requires a bunch of assumptions:
* A server has published an IPv4 address in the public_name even though they're not supposed to, presumably because it's permitted by a future draft.
* public_name is delivered "securely".
* ICANN won't let anyone register a numeric TLD.
* Clients don't parse IPv4 strings into 4 octets, and then try to interpret that as a domain name, resulting in a memory safety violation.

I find these assumptions less than totally obvious, and that makes this text confusing.

Here's some alternative explanatory text.
```suggestion
value when the input is an IPv4 address.) Ignoring IPv4 addresses in `public_name`
allows their use to be defined in a future update to ECH without breaking
compatibility.
```

Reviewer: > Ignoring IPv4 addresses in `public_name` allows their use to be defined in a future update to ECH without breaking compatibility.

I don't think this is true. If that's the goal, we'd need a MUST-level requirement. With a SHOULD, some TLS implementations will ignore it, while others will go ahead and use that ECHConfig and likely get confused.

Author: @bemasc clarifying question: by "ignoring IPv4 addresses in `public_name`," do you mean "ignore ECHConfigs that contain an IPv4 address," or "allow ECHConfigs that contain an IPv4 address (by ignoring the fact that the config contains one)"? (I'm parsing your suggestion as the latter, which seems counter to what this PR is trying to do.)

Reviewer 4: @davidben Perhaps, but that's how I read the intention of the PR nonetheless.

@chris-wood The former.  I updated my comment to clarify.

> to the a reference identity based on the public_name.

Reviewer: s/the a/the/

> to the a reference identity based on the public_name.

Author: ```suggestion
to the reference identity based on the public_name.
```

> server. However, it is not expected to induce security failures. Clients will
> use this name to authenticate the client-facing server and will fail closed if
> the client-facing server cannot produce a certificate bound to the same name.

Reviewer 3: "to the same name" might be tricky - if there is a disagreement about whether the name is an IP or DNS name.

How about:

```suggestion
server. However, it is not expected to induce security failures. Clients will
fail closed if the client-facing server cannot produce a certificate bound
to the a reference identity based on the public_name.
```

Author: Much better!

   allows servers that reject ECH and use ClientHelloOuter to safely ignore any
   early data sent by the client per {{RFC8446}}, Section 4.2.10.

Note that these rules may change in the presence of an application profile

> specifying otherwise.

Reviewer: No objection to the sentence, though I'm unsure what it's trying to do in the context of this PR.

Author: I threw this in to allow clients to omit the public name in CHO.

Reviewer 5: If someone is allowed to change these rules, then how can they really be MUSTs?

> standard specifying otherwise.

Reviewer 3: ```suggestion
specifying otherwise.
```
